File: ts/src/validate.ts
/** Light validation of params against the compact schema (SPEC §4.1.1), producing teaching errors. */
import { fix, YeaError } from './errors.js';
import type { Fix, ParamSchema } from './types.js';

export function distance(a: string, b: string): number {
  const d = Array.from({ length: a.length + 1 }, (_, i) => [
    i,
    ...Array(b.length).fill(0),
  ]);

  for (let j = 1; j <= b.length; j++) {
    d[0][j] = j;
  }

  for (let i = 1; i <= a.length; i++) {
    for (let j = 1; j <= b.length; j++) {
      d[i][j] = Math.min(
        d[i - 1][j] + 1,
        d[i][j - 1] + 1,
        d[i - 1][j - 1] + (a[i - 1] === b[j - 1] ? 0 : 1),
      );
    }
  }

  return d[a.length][b.length];
}

export function closest(word: string, options: string[]): string | undefined {
  if (word.length > 64) {
    return undefined; // suggestions are for typos; don't pay O(n·m) on garbage
  }

  let best: string | undefined;
  let bestD = Infinity;

  for (const o of options) {
    const dd = distance(word.toLowerCase(), o.toLowerCase());

    if (dd < bestD) {
      best = o;
      bestD = dd;
    }
  }

  return best !== undefined && bestD <= Math.max(2, Math.floor(word.length / 3))
    ? best
    : undefined;
}

const DATE = /^\d{4}-\d{2}-\d{2}$/;
const DATETIME =
  /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}(:\d{2}(\.\d+)?)?(Z|[+-]\d{2}:\d{2})$/;

/** Named scalar types: a test for the value, and how to describe what was wanted. */
const SCALARS = new Map<
  string,
  { test: (v: unknown) => boolean; want: string }
>([
  ['string', { test: (v) => typeof v === 'string', want: 'a string' }],
  ['int', { test: Number.isInteger, want: 'an integer' }],
  ['number', { test: (v) => typeof v === 'number', want: 'a number' }],
  ['bool', { test: (v) => typeof v === 'boolean', want: 'true or false' }],
  [
    'date',
    {
      test: (v) => typeof v === 'string' && DATE.test(v),
      want: 'a date like 2026-09-24',
    },
  ],
  [
    'datetime',
    {
      test: (v) => typeof v === 'string' && DATETIME.test(v),
      want: 'an ISO datetime like 2026-09-24T15:00:00Z',
    },
  ],
  ['any', { test: () => true, want: 'anything' }],
]);

/** What `v` should have been for `type` (e.g. "an integer"), or null if it conforms. Unknown types pass. */
function checkType(type: string, v: unknown): string | null {
  const base = type.split(' — ')[0].trim();

  if (base.endsWith('[]')) {
    const item = base.slice(0, -2);

    return Array.isArray(v) && v.every((x) => !checkType(item, x))
      ? null
      : `an array of ${item}`;
  }

  const scalar = SCALARS.get(base);

  if (scalar) {
    return scalar.test(v) ? null : scalar.want;
  }

  if (base.includes('|')) {
    const opts = base.split('|');

    return opts.includes(String(v)) ? null : `one of ${opts.join(', ')}`;
  }

  return null;
}

const isPlainObject = (v: unknown): v is Record<string, unknown> =>
  typeof v === 'object' && v !== null && !Array.isArray(v);

/** Problems inside a nested object, each already prefixed with `path`. */
function nestedProblems(
  schema: ParamSchema | undefined,
  value: Record<string, unknown>,
  path: string,
): string[] {
  try {
    validateParams(schema, value, path);

    return [];
  } catch (e) {
    return [(e as Error).message];
  }
}

function arrayOfObjectsProblems(
  label: string,
  schema: ParamSchema | undefined,
  v: unknown,
): string[] {
  if (!Array.isArray(v)) {
    return [`\`${label}\` must be an array of objects`];
  }

  return v.flatMap((item, i) =>
    isPlainObject(item)
      ? nestedProblems(schema, item, `${label}.${i}.`)
      : [`\`${label}.${i}\` must be an object`],
  );
}

/** Problems with one declared param; `label` is its full dotted path. */
function fieldProblems(
  label: string,
  type: ParamSchema[string],
  v: unknown,
  optional: boolean,
): string[] {
  if (v === undefined || v === null) {
    return optional
      ? []
      : [
          `missing \`${label}\` (${typeof type === 'string' ? type : 'object'})`,
        ];
  }

  if (typeof type === 'string') {
    const want = checkType(type, v);

    return want
      ? [`\`${label}\` must be ${want} (got ${JSON.stringify(v)})`]
      : [];
  }

  if (Array.isArray(type)) {
    return arrayOfObjectsProblems(label, type[0], v);
  }

  if (!isPlainObject(v)) {
    return [`\`${label}\` must be an object`];
  }

  return nestedProblems(type, v, `${label}.`);
}

/** Params the schema doesn't declare, with a rename fix when one is a likely typo. */
function unknownParams(
  params: Record<string, unknown>,
  names: string[],
  path: string,
): { problems: string[]; fixes: Fix[] } {
  const problems: string[] = [];
  const fixes: Fix[] = [];

  for (const k of Object.keys(params)) {
    if (names.includes(k)) {
      continue;
    }

    const near = closest(k, names);

    problems.push(`unknown param \`${path}${k}\``);

    if (near && params[near] === undefined) {
      fixes.push(
        fix(`rename \`${k}\` to \`${near}\``, { [k]: null, [near]: params[k] }),
      );
    }
  }

  return { problems, fixes };
}

/** Throws a YeaError(invalid_params) with fixes if params don't match the schema. */
export function validateParams(
  schema: ParamSchema | undefined,
  params: Record<string, unknown>,
  path = '',
): void {
  if (!schema) {
    return;
  }

  const problems: string[] = [];
  const names = Object.keys(schema).map((k) => k.replace(/\?$/, ''));

  for (const [raw, type] of Object.entries(schema)) {
    const name = raw.replace(/\?$/, '');

    problems.push(
      ...fieldProblems(`${path}${name}`, type, params[name], raw.endsWith('?')),
    );
  }

  const unknown = unknownParams(params, names, path);

  problems.push(...unknown.problems);

  if (problems.length) {
    throw new YeaError(
      'invalid_params',
      problems.join('; '),
      unknown.fixes.length ? { fix: unknown.fixes } : {},
    );
  }
}

File: python/src/yea/validate.py
"""Light validation of params against the compact schema (SPEC §4.1.1), with teaching errors.

Mirrors ts/src/validate.ts so both implementations teach the same way.
"""

from __future__ import annotations

import re
from typing import Any

from ._json import compact
from .errors import YeaError, fix

_DATE = re.compile(r"\d{4}-\d{2}-\d{2}")
_DATETIME = re.compile(r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}(:\d{2}(\.\d+)?)?(Z|[+-]\d{2}:\d{2})")


def distance(a: str, b: str) -> int:
    prev = list(range(len(b) + 1))
    for i, ca in enumerate(a, 1):
        cur = [i]
        for j, cb in enumerate(b, 1):
            cur.append(min(prev[j] + 1, cur[j - 1] + 1, prev[j - 1] + (ca != cb)))
        prev = cur
    return prev[-1]


def closest(word: str, options: list[str]) -> str | None:
    if len(word) > 64:
        return None  # suggestions are for typos; don't pay O(n·m) for arbitrary input
    best, best_d = None, None
    for o in options:
        d = distance(word.lower(), o.lower())
        if best_d is None or d < best_d:
            best, best_d = o, d
    return best if best is not None and best_d <= max(2, len(word) // 3) else None


def _is_int(v: Any) -> bool:
    return (type(v) is int) or (isinstance(v, float) and v.is_integer())


def _check_type(type_: str, v: Any) -> str | None:
    base = type_.split(" — ")[0].strip()
    if base.endswith("[]"):
        inner = base[:-2]
        if not isinstance(v, list) or any(_check_type(inner, x) for x in v):
            return f"an array of {inner}"
        return None
    if base == "string":
        return None if isinstance(v, str) else "a string"
    if base == "int":
        return None if _is_int(v) and not isinstance(v, bool) else "an integer"
    if base == "number":
        return None if isinstance(v, (int, float)) and not isinstance(v, bool) else "a number"
    if base == "bool":
        return None if isinstance(v, bool) else "true or false"
    if base == "date":
        return None if isinstance(v, str) and _DATE.fullmatch(v) else "a date like 2026-09-24"
    if base == "datetime":
        return None if isinstance(v, str) and _DATETIME.fullmatch(v) else "an ISO datetime like 2026-09-24T15:00:00Z"
    if "|" in base:
        opts = base.split("|")
        return None if _js_string(v) in opts else f"one of {', '.join(opts)}"
    return None  # "any", or a type this implementation doesn't know


def _js_string(v: Any) -> str:
    if v is True:
        return "true"
    if v is False:
        return "false"
    if v is None:
        return "null"
    return v if isinstance(v, str) else compact(v)


def validate_params(schema: dict | None, params: dict, path: str = "") -> None:
    """Raise YeaError(invalid_params) with fixes if ``params`` don't match ``schema``."""
    if not schema:
        return
    problems: list[str] = []
    fixes: list[dict] = []
    names = [k[:-1] if k.endswith("?") else k for k in schema]
    for raw, type_ in schema.items():
        optional = raw.endswith("?")
        name = raw[:-1] if optional else raw
        v = params.get(name)
        if v is None:
            if not optional:
                problems.append(f"missing `{path}{name}` ({type_ if isinstance(type_, str) else 'object'})")
            continue
        if isinstance(type_, str):
            want = _check_type(type_, v)
            if want:
                problems.append(f"`{path}{name}` must be {want} (got {compact(v)})")
        elif isinstance(type_, list):
            if not isinstance(v, list):
                problems.append(f"`{path}{name}` must be an array of objects")
            else:
                for i, item in enumerate(v):
                    if not isinstance(item, dict):
                        problems.append(f"`{path}{name}.{i}` must be an object")
                        continue
                    try:
                        validate_params(type_[0], item, f"{path}{name}.{i}.")
                    except YeaError as e:
                        problems.append(e.message)
        elif not isinstance(v, dict):
            problems.append(f"`{path}{name}` must be an object")
        else:
            try:
                validate_params(type_, v, f"{path}{name}.")
            except YeaError as e:
                problems.append(e.message)
    for k in params:
        if k in names:
            continue
        near = closest(k, names)
        problems.append(f"unknown param `{path}{k}`")
        if near and params.get(near) is None:
            fixes.append(fix(f"rename `{k}` to `{near}`", {k: None, near: params[k]}))
    if problems:
        raise YeaError("invalid_params", "; ".join(problems), fix=fixes or None)
